docs/ratt-transform/index.md

```diff
@@ -675,84 +675,384 @@ app.use(
 ```json
 {
-  "name": "J.D.",
-  "properties": [
+  "metadata": {
+    "title": {
+      "name": "Data about countries."
+    }
+  },
+  "data": {
+    "countries": [
+      {
+        "id": "nl",
+        "name": "The Netherlands"
+      },
+      {
+        "id": "de",
+        "name": "Germany"
+      }
+    ]
+  }
+}
+```
+
+Since the RATT record abstracts away the syntactic specifics of the source system format, the above example may be read from a JSON source (see [connecting JSON sources](#json)), or from another kind of source.  For example, the RATT record may be read from the following XML source, assuming the contents of the `<root>` tag are selected:
+
+```xml
+<?xml version="1.0"?>
+<root>
+  <metadata>
+    <title>
+      <value>Data about countries.</value>
+    </title>
+  </metadata>
+  <data>
+    <countries>
+      <id>nl</id>
+      <name>The Netherlands</name>
+    </countries>
+    <countries>
+      <id>de</id>
+      <name>Germany</name>
+    </countries>
+  </data>
+</root>
+```
+
+See [the section on connecting XML sources](#xml) for more information.
+
+### Specifying paths (nested keys)
+
+In tabular data, keys (or column names) are singular.  But in tree-shaped data a *path* of the tree can consist of one or more keys that must be traversed in sequence.
+
+Paths are specified as dot-separated sequences of keys, starting at the top-level and ending at the required value.  For the above example, RATT can access the `"name"` key inside the `"title"` key, which itself is nested inside the `"metadata"` key.  This path is expressed in [1].  Notice that the path expressed in [1] is different from the path expressed in [2], which also access the `"name"` key, but nested inside the `"countries"` and then `"data"` keys.
+
+```
+[1] metadata.title.name
+[2] data.countries.name
+```
+
+Path expressions can be used anywhere with singular keys.  For example, we can assert the title of a dataset in the following way:
+
+```ts
+app.use(
+  mw.addQuad(
+    prefix.dataset('my-dataset'),
+    dct.title,
+    mw.toLiteral('metadata.title.value', {language: 'en'})),
+)
+```
+
+This results in the following assertion:
+
+```trig
+dataset:my-dataset dct:title 'Data about countries.'@en.
+```
+
+<h3 id='accessing-lists-by-index'>Accessing lists by index</h3>
+
+Tree-shaped data formats often allow multiple values to be specified in a sequence.  Examples of this are XML tags with the same key that are directly nested under the same parent, and lists in JSON.
+
+RATT is able to access specific elements from such lists based on their *index* or position.  Following the standard practice in Computer Science, RATT refer to the first element in the list as having index 0.  The second element has index 1, etc.
+
+For the above example record, we can assert the first label of the country:
```

Comment:
These examples are not about list elements, but objects in an array. It is important to mention that because later RATT works differently with objects that have arrays as values ( `mw.toLiteral.forEach()` and `mw.toIri.forEach()`) and objects with arrays of objects (`mw.forEach()`).

Reply:
It is indeed good to think about the terminology here: JSON uses "array" for ordered lists. XML is not very explicit about how ordered lists are called, talking about the plural "elements" inside a parent "element".  I'm dropping my use of "sequence" here, because that term is neither used in JSON nor XML.  Let's go with "ordered list" that is more descriptive and less implementation-specific than "array" or "elements".

I agree that the user should be warned that `forEach` behaves differently on lists of objects than on lists of strings or numbers, but that warning should be part of the documentation for `forEach`.

Reply:
Ah, `forEach` follows quickly in the next section I see...

Reply:
`mw.toIri.forEach` and `mw.toLiteral.forEach` are not documented yet...

Reply:
I have added a new section for iterating over lists of primitives.